File: dfs/longest_cycle/longest_cycle.cpp
#include <vector>
using namespace std;

class LongestCycle
{ // LeetCode Q.2360.
private:
    vector<int> graph;
    vector<int> visitOrders;
    vector<int> status; // 1: not searched yet; 2: now under DFS; 3: finished DFS.

    int dfsCycleLen(int node, int visitOrder)
    {
        visitOrders[node] = visitOrder, status[node] = 2;

        int cycleLen = -1; // Default to no cycle.
        if (graph[node] != -1)

        { // Current node has an outgoing node.
            if (status[graph[node]] == 1)
            {
                cycleLen = dfsCycleLen(graph[node], visitOrder + 1);
            }

            if (status[graph[node]] == 2)
            { // Cycle forms.
                cycleLen = visitOrder + 1 - visitOrders[graph[node]];
            }
        }

        status[node] = 3; // Current node finishes DFS.
        return cycleLen;
    }

public:
    int computeLongestCycleLen(vector<int> &edges)
    {
        graph = edges;
        visitOrders.resize(edges.size(), -1); // Default order = -1 (not searched).
        status.resize(edges.size(), 1);       // All nodes aren't searched at the beginning.

        int maxCycleLen = -1;
        for (int outNode = 0; outNode < edges.size(); outNode++)
        {
            int inNode = edges[outNode];
            if (inNode != -1)

            { // Each node has at most 1 outgoing edge.
                int cycleLen = dfsCycleLen(outNode, 1);
                if (cycleLen > maxCycleLen)
                {
                    maxCycleLen = cycleLen;
                }
            }
        }

        return maxCycleLen;
    }
};

File: dfs/longest_cycle/longest_cycle.py

class LongestCycle:  # LeetCode Q.2360.
    def __init__(self):
        self.graph: list[int] = []
        self.visit_orders: list[int] = []
        self.status: list[int] = []  # 1: not searched yet; 2: now under DFS; 3: finished DFS.

    def _dfs_cycle_len(self, node: int, visit_order: int) -> int:
        self.visit_orders[node], self.status[node] = visit_order, 2
        cycle_len = -1  # Default to no cycle.

        if self.graph[node] != -1:  # Node has an outgoing edge.
            if self.status[self.graph[node]] == 1:
                cycle_len = self._dfs_cycle_len(self.graph[node], visit_order + 1)

            if self.status[self.graph[node]] == 2:  # Cycle forms.
                cycle_len = visit_order + 1 - self.visit_orders[self.graph[node]]

        self.status[node] = 3  # Current node finishes DFS.
        return cycle_len

    def compute_longest_cycle_len(self, edges: list[int]) -> int:
        self.graph.clear()  # Reset before search.
        self.visit_orders.clear()
        self.status.clear()

        self.graph.extend(edges)
        self.visit_orders.extend([-1] * len(edges))  # Default order = -1 (not searched).
        self.status.extend([1] * len(edges))  # All nodes aren't searched at the beginning.

        max_cycle_len = -1
        for out_node, in_node in enumerate(edges):  # Each node joins at most 1 cycle.
            if in_node != -1:  # Each node has at most 1 outgoing edge.
                cycle_len = self._dfs_cycle_len(out_node, 1)
                if cycle_len > max_cycle_len:
                    max_cycle_len = cycle_len

        return max_cycle_len
